extensions/amp-a4a/0.1/real-time-config-manager.js

```diff
@@ -0,0 +1,167 @@
+    this.inflateVendorUrls();
+    if (this.calloutUrls.length == 0) {
+      return Promise.resolve();
+    }
+    let rtcTime;
+    return Promise.all(this.calloutUrls.map(url => {
+      const rtcStartTime = Date.now();
+      /** ISSUE: For very short timeouts, and/or for slow devices, the calculated time
+       * that RTC took (rtcTime below) can be longer than the timeout without getting timed-out,
+       * due to the time it takes to set up the timeout promise.
+       * EXAMPLE, let rtcTimeout = 100ms.
+       * Time = 0: const rtcStartTime = 0
+       * Time = 10: call Services.timerFor.timeoutPromise
+       * Time = 25: timeout promise starts the xhr callout, timeout promise sees start time as 25
+       * Time = 120: xhr callout returns, timeoutPromise sees total time as 95ms, so no timeout
+       * Time = 125: rtcTime = 125ms, which is greater than the publisher specified timeout.
+       */
+      return Services.timerFor(this.win).timeoutPromise(
+          rtcTimeout,
+          Services.xhrFor(this.win).fetchJson(
+              url, {credentials: 'include'}).then(res => {
+                // Non-200 status codes are forbidden for RTC.
+                // TODO: Add to fetchResponse the ability to
+                // check for redirects as well.
+                if (res.status != 200) {
+                  rtcTime = Date.now() - rtcStartTime;
+                  return {rtcTime, error: 'Non-200 Status', hostname};
+                }
+                return res.text().then(text => {
+                  rtcTime = Date.now() - rtcStartTime;
+                  const hostname = parseUrl(url).hostname;
+                  // An empty text response is fine, just means
+                  // we have nothing to merge.
+                  if (!text) {
+                    return {rtcTime, hostname};
+                  }
+                  const rtcResponse = tryParseJson(text);
+                  return rtcResponse ? {rtcResponse, rtcTime, hostname} :
+                  {rtcTime, hostname, error: 'Unparsable JSON'};
+                });
+              })).catch(err => {
+                const hostname = parseUrl(url).hostname;
+                const rtcTime = Date.now() - rtcStartTime;
+                return {error: err, rtcTime, hostname};
+              });
+    }));
+  }
+
+  /**
+   * Attempts to parse the publisher-defined RTC config off the amp-ad
+   * element, then validates that the rtcConfig exists, and contains
+   * an entry for either vendor URLs, or publisher-defined URLs. If the
+   * config contains an entry for timeoutMillis, validates that it is a
+   * number, or converts to a number if number-like, otherwise overwrites
+   * with the default.
+   * @return {!boolean}
+   */
+  validateRtcConfig() {
+    let rtcConfig = tryParseJson(
+        this.element.getAttribute('prerequest-callouts'));
+    if (!rtcConfig) {
+      return false;
+    }
+    if (!((rtcConfig['vendors'] && isObject(rtcConfig['vendors']) &&
```

Comment:
I don't think we need to check length... if they're both empty the effect would be no RTC, right?
try {
   user().assert(rtcConfig['vendors'] || rtcConfig['urls], 'RTC must defined vendors or urls...');
   user().assert(!rtcConfig['vendors'] || isObject(rtcConfig['vendors']), 'RTC invalid vendors... ');
   user().assert(!rtcConfig['urls'] || isArray(rtcConfig['urls']), 'RTC invalid urls...');
} catch (err) {
   return false;
}